modules/utils/graph/Redif.cpp

```diff
@@ -0,0 +1,325 @@
+#include "Redif.hpp"
+
+
+namespace metric
+{
+	template <typename recType, class Metric, typename Tv>
+	Redif<recType, Metric, Tv>::Redif(std::vector<std::vector<Tv>> trainData, size_t nNeighbors /* = 10 */, size_t nIter /* = 15 */)
+	{
+		this->nNeighbors = nNeighbors;
+		xTrain = blaze::DynamicMatrix<Tv>(trainData.size(), trainData[0].size());
+		for (size_t i = 0; i < xTrain.rows(); i++)
+			for (size_t j = 0; j < xTrain.columns(); j++)
+				xTrain(i, j) = trainData[i][j];
+		blaze::DynamicMatrix<Tv> trainDe(trainData.size(), trainData[0].size());
+		trainModel(trainDe, nIter);
+	}
+
+
+	template <typename recType, class Metric, typename Tv>
+	void Redif<recType, Metric, Tv>::trainModel(blaze::DynamicMatrix<Tv> & xTrainDe, size_t nIter)
+	{
+		xTrainDe = xTrain;
+		size_t n = xTrainDe.rows();
+
+		/// Iterative backward Diffusion process
+		for (size_t i = 0; i < nIter; i++)
+		{
+			blaze::DynamicMatrix<Tv> localDistMat = getLocalDistMatrix(xTrainDe);
+			blaze::DynamicMatrix<Tv> graphLap = calcWeightedGraphLaplacian(localDistMat);
+			LArray.push_back(graphLap);
+			blaze::DynamicMatrix<Tv> D(n, n);
+			for (size_t i = 0; i < D.rows(); i++)
+				D(i, i) = graphLap(i, i);
+			blaze::DynamicMatrix<Tv> invLeft(n, n);
+			invLeft = blaze::inv(D + 0.25*graphLap);
+			xTrainDe = invLeft * (D*xTrainDe);
+
+			//cout << endl << endl << endl << endl << endl << endl;
+			//cout << i << "-th" << endl;
+			//for (size_t j = 0; j < 10; j++)
+			//	cout << graphLap(0, j) << "\t";
+		}
+
+	}
+
+	template <typename recType, class Metric, typename Tv>
+	blaze::DynamicMatrix<Tv> Redif<recType, Metric, Tv>::getLocalDistMatrix(blaze::DynamicMatrix<Tv> dataSample)
+	{
+		size_t n = dataSample.rows();
+
+		blaze::DynamicMatrix<size_t> ixx(nNeighbors, n);
+		for (size_t i = 0; i < nNeighbors; i++)
+			for (size_t j = 0; j < n; j++)
+				ixx(i, j) = j;
+
+		blaze::DynamicMatrix<Tv> dist(n, n);
+		for (size_t i = 0; i < n; i++)
+			for (size_t j = 0; j < n; j++)
+			{
+				if (i == j)
+				{
+					dist(i, j) = 0;
+					continue;
+				}
+				blaze::DynamicVector<Tv, rowVector> rowi = row(dataSample, i);
+				blaze::DynamicVector<Tv, rowVector> rowj = row(dataSample, j);
+
+
+				std::vector<Tv> veci, vecj;
+				for (size_t k = 0; k < rowi.size(); ++k)
+				{
+					veci.push_back(rowi[k]);
+					vecj.push_back(rowj[k]);
+				}				
+
+				Metric metric;

```

Comment:
this code makes it really hard to provide custom metric. can we make it a method's argument instead? Also if this is the only place where we use metric we could move template parameter from class here.